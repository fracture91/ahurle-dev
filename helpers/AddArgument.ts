// https://stackoverflow.com/a/60324381

// some utility types for working with tuples
type Cons<H, T extends readonly any[]> = ((
  head: H,
  ...tail: T
) => void) extends (...cons: infer R) => void
  ? R
  : never

type Push<T extends readonly any[], V> = T extends any
  ? Cons<void, T> extends infer U
    ? { [K in keyof U]: K extends keyof T ? T[K] : V }
    : never
  : never

// final type you need
export type AddArgument<F, Arg> = F extends (...args: infer PrevArgs) => infer R
  ? (...args: Push<PrevArgs, Arg>) => R
  : never
