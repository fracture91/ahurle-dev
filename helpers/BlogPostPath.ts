/* eslint-disable max-classes-per-file */
// It's a lot easier to read when these two classes are together
import glob from "glob"
import Path from "path"

const MD_EXT = ".mdx"
const PAGES_DIR_FROM_ROOT = "./pages"
const BLOG_DIR_FROM_ROOT = `${PAGES_DIR_FROM_ROOT}/blog`

export class PagePath {
  protected pathFromRoot: string

  public constructor({ pathFromRoot }: { pathFromRoot: string }) {
    this.pathFromRoot = Path.normalize(pathFromRoot)
    if (!this.pathFromRoot.endsWith(MD_EXT)) {
      this.pathFromRoot += MD_EXT
    }
  }

  get pathFromPagesDir(): string {
    return this.pathFromRoot.replace(/^pages\//, "")
  }

  get urlPath(): string {
    return this.pathFromPagesDir.replace(/\..*$/, "")
  }

  get slug(): string {
    const slug = Path.basename(
      this.pathFromRoot,
      Path.extname(this.pathFromRoot)
    )
    if (!slug) throw new Error(`Invalid page path: ${this.pathFromRoot}`)
    return slug
  }

  equals(other: PagePath): boolean {
    return this.pathFromRoot === other.pathFromRoot
  }

  myClass<T extends typeof PagePath>(): T {
    return (this.constructor as unknown) as T
  }

  glob<T extends typeof PagePath>(this: InstanceType<T>): InstanceType<T>[] {
    return glob
      .sync(this.pathFromRoot)
      .map((p) => this.myClass<T>().relativeToRoot<T>(p))
  }

  static relativeToRoot<T extends typeof PagePath>(
    this: T,
    pathFromRoot: string
  ): InstanceType<T> {
    return new this({ pathFromRoot }) as InstanceType<T>
  }

  static relativeToPagesDir<T extends typeof PagePath>(
    this: T,
    pathFromPagesDir: string
  ): InstanceType<T> {
    return new this({
      pathFromRoot: `${PAGES_DIR_FROM_ROOT}/${pathFromPagesDir}`,
    }) as InstanceType<T>
  }
}

export class BlogPostPath extends PagePath {
  get pathFromBlogDir(): string {
    return this.pathFromPagesDir.replace(/^blog\//, "")
  }

  get urlPath(): string {
    return `blog/${this.slug}`
  }

  // I wish I could narrow the types here, but that prevents overriding
  glob<T extends typeof PagePath>(this: InstanceType<T>): InstanceType<T>[] {
    return (
      super.glob
        // typescript gets confused about the type of `this` without `call` :(
        .call<InstanceType<T>, never, InstanceType<T>[]>(this)
        // the point of this override is to exclude the index page - it is not itself a "blog post"
        .filter((p) => !p.equals(BlogPostPath.INDEX_PAGE))
    )
  }

  // even with the above trickery you still need to provide a type parameter
  // to get a BlogPostPath instead of a PagePath, so make usage a little easier... sigh
  globWithoutBlogIndex(): BlogPostPath[] {
    return this.glob<typeof BlogPostPath>()
  }

  static fromSlug(slug: string): BlogPostPath {
    return new BlogPostPath({
      pathFromRoot: `${BLOG_DIR_FROM_ROOT}/${slug}`,
    })
  }

  static relativeToBlogDir(pathFromBlogDir: string): BlogPostPath {
    return new BlogPostPath({
      pathFromRoot: `${BLOG_DIR_FROM_ROOT}/${pathFromBlogDir}`,
    })
  }

  static INDEX_PAGE = BlogPostPath.relativeToBlogDir("index.mdx")
}
