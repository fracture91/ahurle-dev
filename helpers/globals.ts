const prodUrl = "https://ahurle.dev"

export const yourName = "Andrew Hurle"
export const siteName = "ahurle.dev"
export const siteDescription = "Dev Blog"
export const siteCreationDate = "January 15, 2021 20:00:00 GMT"
// export const twitterHandle = '@';
export const email = "me@ahurle.dev"
export const url =
  // eslint-disable-next-line no-nested-ternary
  process.env.VERCEL_ENV &&
  process.env.VERCEL_ENV !== "production" &&
  process.env.VERCEL_URL
    ? `https://${process.env.VERCEL_URL}`
    : process.env.NODE_ENV === "production"
    ? prodUrl
    : "http://localhost:3000"
export const goatCounterId = "ahurle-dev"
export const tinyLetterUsername = "ahurle"
