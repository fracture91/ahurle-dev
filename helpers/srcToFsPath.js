// @ts-check
/**
 * Supports three ways to specify the source:
 *
 * ```
 * // using next-images webpack loader
 * import pancake from "public/img/pancake.jpg"
 *
 * src = pancake // "/_next/static/images/pancakes-1234abcd.jpg"
 * src = "img/pancake.jpg" // optional leading slash, below as well
 * src = "public/img/pancake.jpg"
 * ```
 *
 * Converts each case into a file path relative to project root.
 *
 * @param {string} src
 * @returns {string}
 */

module.exports = (src) => {
  let done = false
  const result = src.replace(/^\//, "").replace(/^_next\//, () => {
    done = true
    return ".next/"
  })
  if (done) return result

  return result.replace(/^public\/|^/, "public/")
}
