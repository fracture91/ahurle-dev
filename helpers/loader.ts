import sizeOf from "image-size"
import type { MDXModule } from "types/mdx.d"
import { globals } from "./globals"
import {
  BlogMeta,
  RawBlogMetaInput,
  RawBlogMetaOutput,
  RawBlogMetaSchema,
  BannerPhoto,
  RawBannerPhoto,
} from "./schema"
import { BlogPostPath } from "./BlogPostPath"

export interface LayoutProps {
  path: string
  excerpt: string
  meta: RawBlogMetaInput
}

export type MDXBlogModule = LayoutProps & MDXModule

type ModuleAndPath = { path: BlogPostPath; module: MDXBlogModule }

export const loadRawBlogPost = async (
  path: BlogPostPath
): Promise<ModuleAndPath> => {
  // important: need "../pages/blog/" here explicitly to help out webpack
  const module = await import(`../pages/blog/${path.pathFromBlogDir}`)
  if (
    typeof module.path !== "string" ||
    !BlogPostPath.relativeToRoot(module.path).equals(path)
  ) {
    throw new Error("Path mismatch - a bug or missing remark plugin?")
  }
  return { path, module }
}

const hasSize = (raw: RawBannerPhoto): raw is BannerPhoto =>
  !!(raw.width && raw.height)

/**
 * Supports three ways to specify the source:
 *
 * ```
 * // using next-images webpack loader
 * import pancake from "public/img/pancake.jpg"
 *
 * src = pancake // "/_next/static/images/pancakes-1234abcd.jpg"
 * src = "img/pancake.jpg" // optional leading slash, below as well
 * src = "public/img/pancake.jpg"
 * ```
 *
 * Converts each case into a file path relative to project root.
 */
export const srcToFsPath = (src: string): string => {
  let done = false
  const result = src.replace(/^\//, "").replace(/^_next\//, () => {
    done = true
    return ".next/"
  })
  if (done) return result

  return result.replace(/^public\/|^/, "public/")
}

export const processRawMeta = async ({
  module,
  path,
}: {
  module: LayoutProps
  path: BlogPostPath
}): Promise<BlogMeta> => {
  try {
    const { urlPath } = path
    const raw: RawBlogMetaOutput = RawBlogMetaSchema.parse(module.meta)

    let bannerPhoto: BannerPhoto | undefined
    if (raw.bannerPhoto) {
      if (hasSize(raw.bannerPhoto)) {
        bannerPhoto = raw.bannerPhoto
      } else {
        const src = srcToFsPath(raw.bannerPhoto.src)
        const { width, height } = await sizeOf(src)
        if (!width || !height) {
          throw new Error(`Could not get image size: ${src}`)
        }
        bannerPhoto = { ...raw.bannerPhoto, width, height }
      }
    }

    return {
      ...raw,
      urlPath,
      description: raw.description || module.excerpt,
      slug: path.slug,
      canonicalUrl: new URL(raw.canonicalUrl || urlPath, globals.url).href,
      bannerPhoto,
    }
  } catch (e) {
    throw new Error(
      `Error parsing meta for blog post ${path.pathFromBlogDir}: ${e.message}`
    )
  }
}

export const loadRawBlogPosts = (
  paths: BlogPostPath[]
): Promise<ModuleAndPath>[] => paths.map((path) => loadRawBlogPost(path))

export const loadBlogMeta = async (path: BlogPostPath): Promise<BlogMeta> => {
  const mp = await loadRawBlogPost(path)
  return processRawMeta(mp)
}

export interface MetaAndContent<Published extends boolean = boolean> {
  meta: BlogMeta<Published>
  content: MDXBlogModule["default"]
}

const isPublished = (meta: BlogMeta): meta is BlogMeta<true> => meta.published
const isMetaAndContentPublished = (
  mc: MetaAndContent
): mc is MetaAndContent<true> => isPublished(mc.meta)

export const loadPublishedBlogs = async (): Promise<MetaAndContent<true>[]> => {
  const paths = BlogPostPath.fromSlug("*").glob()
  const promises = loadRawBlogPosts(paths).map((p) =>
    p.then((mp) =>
      processRawMeta(mp).then((meta) => ({ meta, content: mp.module.default }))
    )
  )
  return (await Promise.all(promises))
    .filter(isMetaAndContentPublished)
    .sort(({ meta: a }, { meta: b }) => b.datePublished - a.datePublished)
}
