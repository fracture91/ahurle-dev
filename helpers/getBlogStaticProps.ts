import { GetStaticProps } from "next"
import removeUndefined from "rundef"
import { BlogMeta } from "@/helpers/schema"
import { BlogPostPath } from "@/helpers/BlogPostPath"
import { LayoutProps, processRawMeta } from "@/helpers/loader"
import type { AddArgument } from "@/helpers/AddArgument"

export type BlogStaticProps = { processedMeta: BlogMeta }

export const getBlogStaticProps: AddArgument<
  GetStaticProps<BlogStaticProps>,
  LayoutProps
> = async (_context, layoutProps) => {
  const { meta, path: rawPath } = layoutProps
  if (!meta.published) {
    return { notFound: true }
  }
  const path = BlogPostPath.relativeToRoot(rawPath)
  return processRawMeta({ module: layoutProps, path }).then(
    (processedMeta) => ({
      props: {
        // next doesn't like it when you have "undefined" because that can't serialize to JSON
        processedMeta: removeUndefined(processedMeta, false, true),
      },
    })
  )
}
