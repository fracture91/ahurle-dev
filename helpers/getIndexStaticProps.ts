import { GetStaticProps, GetStaticPropsResult } from "next"
import removeUndefined from "rundef"
import { BlogMeta } from "@/helpers/schema"
import { generateRSS } from "@/helpers/rssUtil"
import { PageLayoutProps, loadPublishedBlogs } from "@/helpers/loader"
import {
  getPageStaticProps,
  PageStaticProps,
} from "@/helpers/getPageStaticProps"
import type { AddArgument } from "@/helpers/AddArgument"

export type IndexStaticProps = PageStaticProps & { posts: BlogMeta<true>[] }

const hasProps = <T extends GetStaticPropsResult<PageStaticProps>>(
  t: T
): t is T & { props: PageStaticProps } =>
  Object.prototype.hasOwnProperty.call(t, "props")

export const getIndexStaticProps: AddArgument<
  GetStaticProps<IndexStaticProps>,
  PageLayoutProps
> = async (context, layoutProps) => {
  const pageStaticProps = await getPageStaticProps(context, layoutProps)
  if (!hasProps(pageStaticProps)) return pageStaticProps
  const posts = await loadPublishedBlogs()

  /**
   * As written, the dev server will prevent navigation upon writing to the /public dir.
   * Ideally, this step would be a postbuild script instead.  Or even more ideally, we
   * could have a `pages/rss.xml.tsx` containing a `getStaticProps` to generate the file.
   *
   * It's possible to do this with `getServerSideProps`, but that would break `next export`:
   * https://github.com/vercel/next.js/discussions/12403#discussioncomment-10225
   *
   * It's also possible to execute a plain old node script with node/ts-node/babel-node,
   * but I want to execute code that has been through the next.js build pipeline and I
   * don't want to set up an entire parallel build system for one script:
   * https://www.joshwcomeau.com/blog/how-i-built-my-blog/#build-helpers
   * https://leerob.io/blog/nextjs-sitemap-robots#dynamic-sitemaps
   * https://github.com/vercel/next.js/pull/15276#issuecomment-660336199
   *
   * I tried this hack to build a script through next.js, but I gave up on it.
   * Might be possible but not worth the effort.
   * https://dev.to/nalanj/adding-scripts-to-next-js-n7i
   * https://github.com/fracture91/ahurle-dev/commit/3000dc54f07550aeee6665f2e5d84b61a2d268bb
   *
   * So... let's stick with this :/
   */
  // comment out to turn off RSS generation during build step.
  await generateRSS(posts)

  return {
    ...pageStaticProps,
    props: {
      ...pageStaticProps.props,
      posts: posts.map(({ meta }) => removeUndefined(meta, false, true)),
    },
  }
}
