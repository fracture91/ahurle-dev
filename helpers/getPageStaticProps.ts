import { GetStaticProps } from "next"
import removeUndefined from "rundef"
import { PageMeta } from "@/helpers/schema"
import { PagePath } from "@/helpers/BlogPostPath"
import { PageLayoutProps, processRawPageMeta } from "@/helpers/loader"
import type { AddArgument } from "@/helpers/AddArgument"

export type PageStaticProps = { processedMeta: PageMeta }

export const getPageStaticProps: AddArgument<
  GetStaticProps<PageStaticProps>,
  PageLayoutProps
> = async (_context, layoutProps) => {
  const { path: rawPath } = layoutProps
  const path = PagePath.relativeToRoot(rawPath)
  return processRawPageMeta({ module: layoutProps, path }).then(
    (processedMeta) => ({
      props: {
        // next doesn't like it when you have "undefined" because that can't serialize to JSON
        processedMeta: removeUndefined(processedMeta, false, true),
      },
    })
  )
}
