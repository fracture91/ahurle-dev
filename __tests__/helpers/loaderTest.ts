import { BlogPostPath } from "helpers/BlogPostPath"
import * as subject from "helpers/loader"
import srcToFsPath from "helpers/srcToFsPath"
import * as examplePost from "./examplePost.mdx"

describe("loader", () => {
  describe("srcToFsPath", () => {
    ;[
      // normal cases
      ["/_next/static/images/pancake.jpg", ".next/static/images/pancake.jpg"],
      ["_next/static/images/pancake.jpg", ".next/static/images/pancake.jpg"],
      ["img/pancake.jpg", "public/img/pancake.jpg"],
      ["/img/pancake.jpg", "public/img/pancake.jpg"],
      ["public/img/pancake.jpg", "public/img/pancake.jpg"],
      ["/public/img/pancake.jpg", "public/img/pancake.jpg"],
      // weird cases
      ["publicasdf/whatever", "public/publicasdf/whatever"],
      ["_thing/whatever", "public/_thing/whatever"],
      ["/_nextasdf/whatever", "public/_nextasdf/whatever"],
    ].forEach(([input, expected]) => {
      it(`transforms ${input} to ${expected}`, () => {
        expect(srcToFsPath(input)).toEqual(expected)
      })
    })
  })

  describe("processRawMeta", () => {
    let path: BlogPostPath
    let module: subject.LayoutProps
    beforeEach(() => {
      jest
        .useFakeTimers("modern")
        .setSystemTime(new Date("1970-01-02").getTime())

      path = BlogPostPath.relativeToBlogDir("whatever")
      module = {
        meta: { title: "A Title" },
        path: path.pathFromBlogDir,
        excerpt: "wee",
      }
    })

    afterEach(() => {
      jest.useRealTimers()
    })

    it("works for example post", async () => {
      expect(
        await subject.processRawMeta({
          module: examplePost as subject.MDXBlogModule,
          path,
        })
      ).toMatchInlineSnapshot(`
        Object {
          "bannerPhoto": undefined,
          "canonicalUrl": "http://localhost:3000/blog/whatever",
          "datePublished": 86400000,
          "description": "Hello world! This is an example blog post. The excerpt should end after this paragraph.",
          "published": false,
          "slug": "whatever",
          "tags": Array [],
          "title": "An Example Post",
          "urlPath": "blog/whatever",
        }
      `)
    })

    it("works for the minimum required meta", async () => {
      await expect(subject.processRawMeta({ module, path })).resolves
        .toMatchInlineSnapshot(`
              Object {
                "bannerPhoto": undefined,
                "canonicalUrl": "http://localhost:3000/blog/whatever",
                "datePublished": 86400000,
                "description": "wee",
                "published": false,
                "slug": "whatever",
                "tags": Array [],
                "title": "A Title",
                "urlPath": "blog/whatever",
              }
            `)
    })

    it("fails when title is missing", async () => {
      module.meta.title = ""
      await expect(
        subject.processRawMeta({ module, path })
      ).rejects.toBeInstanceOf(Error)
    })

    it("fails when published is true due to missing attrs", async () => {
      module.meta.published = true
      await expect(
        subject.processRawMeta({ module, path })
      ).rejects.toBeInstanceOf(Error)
    })

    it("succeeds when published is true and all required attrs present", async () => {
      module.meta = {
        published: true,
        title: "A Title",
        datePublished: 1234,
        bannerPhoto: {
          src: "img/pancakes.jpeg",
          alt: "alt",
        },
      }
      await expect(subject.processRawMeta({ module, path })).resolves
        .toMatchInlineSnapshot(`
              Object {
                "bannerPhoto": Object {
                  "alt": "alt",
                  "height": 452,
                  "src": "img/pancakes.jpeg",
                  "width": 900,
                },
                "canonicalUrl": "http://localhost:3000/blog/whatever",
                "datePublished": 1234,
                "description": "wee",
                "published": true,
                "slug": "whatever",
                "tags": Array [],
                "title": "A Title",
                "urlPath": "blog/whatever",
              }
            `)
    })

    it("allows specifying a description in place of the excerpt", async () => {
      module.meta.subtitle = "whatever"
      module.meta.description = "my manual description"
      await expect(
        subject.processRawMeta({ module, path }).then((m) => m.description)
      ).resolves.toEqual(module.meta.description)
    })

    it("falls back to subtitle in place of the excerpt", async () => {
      module.meta.subtitle = "my subtitle"
      await expect(
        subject.processRawMeta({ module, path }).then((m) => m.description)
      ).resolves.toEqual(module.meta.subtitle)
    })

    it("passes through bannerPhoto untouched when sizes are provided", async () => {
      module.meta.bannerPhoto = {
        src: "img/idontexist.jpeg",
        width: 123,
        height: 123,
        alt: "alt",
      }
      await expect(
        subject.processRawMeta({ module, path }).then((m) => m.bannerPhoto)
      ).resolves.toMatchInlineSnapshot(`
              Object {
                "alt": "alt",
                "height": 123,
                "src": "img/idontexist.jpeg",
                "width": 123,
              }
            `)
    })
  })

  describe("mdx remark/rehype/webpack behavior", () => {
    it("Adds excerpt, path exports to mdx files", () => {
      expect(examplePost).toMatchInlineSnapshot(`
        Object {
          "default": [Function],
          "excerpt": "Hello world! This is an example blog post. The excerpt should end after this paragraph.",
          "frontMatter": Object {},
          "meta": Object {
            "title": "An Example Post",
          },
          "path": "__tests__/helpers/examplePost.mdx",
        }
      `)
    })

    it("highlights code blocks, adds sizes to images, and otherwise parses things correctly", () => {
      expect(examplePost.default({})).toMatchSnapshot()
    })
  })
})
