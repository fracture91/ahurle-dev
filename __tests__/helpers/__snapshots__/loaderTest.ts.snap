// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader mdx remark/rehype/webpack behavior highlights code blocks, adds sizes to images, and otherwise parses things correctly 1`] = `
<MDXCreateElement
  excerpt="Hello world! This is an example blog post. The excerpt should end after this paragraph."
  frontMatter={Object {}}
  mdxType="wrapper"
  meta={
    Object {
      "title": "An Example Post",
    }
  }
  originalType="wrapper"
  path="__tests__/helpers/examplePost.mdx"
  readingTime={
    Object {
      "minutes": 0.815,
      "text": "1 min read",
      "time": 48900,
      "words": 163,
    }
  }
>
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    Hello world!
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    This is an example blog post.
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    The excerpt should end after this paragraph.
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="blockquote"
    originalType="blockquote"
  >
    <MDXCreateElement
      mdxType="p"
      originalType="p"
      parentName="blockquote"
    >
      a blockquote
    </MDXCreateElement>
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    typescript code block:
  </MDXCreateElement>
  <MDXCreateElement
    className="language-ts"
    mdxType="pre"
    originalType="pre"
  >
    <MDXCreateElement
      className="language-ts"
      mdxType="code"
      originalType="code"
      parentName="pre"
    >
      <MDXCreateElement
        className="token keyword"
        mdxType="span"
        originalType="span"
        parentName="code"
      >
        const
      </MDXCreateElement>
       a 
      <MDXCreateElement
        className="token operator"
        mdxType="span"
        originalType="span"
        parentName="code"
      >
        =
      </MDXCreateElement>
       
      <MDXCreateElement
        className="token number"
        mdxType="span"
        originalType="span"
        parentName="code"
      >
        1
      </MDXCreateElement>
      

    </MDXCreateElement>
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    bash code block:
  </MDXCreateElement>
  <MDXCreateElement
    className="language-bash"
    mdxType="pre"
    originalType="pre"
  >
    <MDXCreateElement
      className="language-bash"
      mdxType="code"
      originalType="code"
      parentName="pre"
    >
      <MDXCreateElement
        className="token builtin class-name"
        mdxType="span"
        originalType="span"
        parentName="code"
      >
        echo
      </MDXCreateElement>
       
      <MDXCreateElement
        className="token string"
        mdxType="span"
        originalType="span"
        parentName="code"
      >
        "hello world"
      </MDXCreateElement>
      

    </MDXCreateElement>
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    plaintext code block:
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="pre"
    originalType="pre"
  >
    <MDXCreateElement
      mdxType="code"
      originalType="code"
      parentName="pre"
    >
      const a = "hello world"

    </MDXCreateElement>
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    Some text,

    <MDXCreateElement
      mdxType="inlineCode"
      originalType="inlineCode"
      parentName="p"
    >
      inline code
    </MDXCreateElement>
    ,

    <MDXCreateElement
      mdxType="strong"
      originalType="strong"
      parentName="p"
    >
      bold
    </MDXCreateElement>
    ,

    <MDXCreateElement
      mdxType="em"
      originalType="em"
      parentName="p"
    >
      italic
    </MDXCreateElement>
    ,

    <MDXCreateElement
      mdxType="del"
      originalType="del"
      parentName="p"
    >
      strikethrough
    </MDXCreateElement>
    ,

    <MDXCreateElement
      href="http://google.com"
      mdxType="a"
      originalType="a"
      parentName="p"
    >
      a link
    </MDXCreateElement>
    ,

    <MDXCreateElement
      href="http://google.com"
      mdxType="a"
      originalType="a"
      parentName="p"
      title="a title"
    >
      a link with title
    </MDXCreateElement>
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="ul"
    originalType="ul"
  >
    <MDXCreateElement
      mdxType="li"
      originalType="li"
      parentName="ul"
    >
      ul item 1
    </MDXCreateElement>
    <MDXCreateElement
      mdxType="li"
      originalType="li"
      parentName="ul"
    >
      ul item 2
    </MDXCreateElement>
    <MDXCreateElement
      mdxType="li"
      originalType="li"
      parentName="ul"
    >
      ul item 3
    </MDXCreateElement>
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="ol"
    originalType="ol"
  >
    <MDXCreateElement
      mdxType="li"
      originalType="li"
      parentName="ol"
    >
      ol item 1
    </MDXCreateElement>
    <MDXCreateElement
      mdxType="li"
      originalType="li"
      parentName="ol"
    >
      ol item 2
    </MDXCreateElement>
    <MDXCreateElement
      mdxType="li"
      originalType="li"
      parentName="ol"
    >
      ol item 3
    </MDXCreateElement>
  </MDXCreateElement>
  <MDXCreateElement
    id="heading-1"
    mdxType="h2"
    originalType="h2"
  >
    <MDXCreateElement
      aria-hidden="true"
      href="#heading-1"
      mdxType="a"
      originalType="a"
      parentName="h2"
      tabIndex={-1}
    >
      <MDXCreateElement
        className="icon icon-link"
        mdxType="span"
        originalType="span"
        parentName="a"
      />
    </MDXCreateElement>
    Heading 1
  </MDXCreateElement>
  <MDXCreateElement
    id="heading-2"
    mdxType="h3"
    originalType="h3"
  >
    <MDXCreateElement
      aria-hidden="true"
      href="#heading-2"
      mdxType="a"
      originalType="a"
      parentName="h3"
      tabIndex={-1}
    >
      <MDXCreateElement
        className="icon icon-link"
        mdxType="span"
        originalType="span"
        parentName="a"
      />
    </MDXCreateElement>
    Heading 2
  </MDXCreateElement>
  <MDXCreateElement
    id="heading-3"
    mdxType="h4"
    originalType="h4"
  >
    <MDXCreateElement
      aria-hidden="true"
      href="#heading-3"
      mdxType="a"
      originalType="a"
      parentName="h4"
      tabIndex={-1}
    >
      <MDXCreateElement
        className="icon icon-link"
        mdxType="span"
        originalType="span"
        parentName="a"
      />
    </MDXCreateElement>
    Heading 3
  </MDXCreateElement>
  <MDXCreateElement
    id="heading-4"
    mdxType="h5"
    originalType="h5"
  >
    <MDXCreateElement
      aria-hidden="true"
      href="#heading-4"
      mdxType="a"
      originalType="a"
      parentName="h5"
      tabIndex={-1}
    >
      <MDXCreateElement
        className="icon icon-link"
        mdxType="span"
        originalType="span"
        parentName="a"
      />
    </MDXCreateElement>
    Heading 4
  </MDXCreateElement>
  <MDXCreateElement
    id="heading-5"
    mdxType="h6"
    originalType="h6"
  >
    <MDXCreateElement
      aria-hidden="true"
      href="#heading-5"
      mdxType="a"
      originalType="a"
      parentName="h6"
      tabIndex={-1}
    >
      <MDXCreateElement
        className="icon icon-link"
        mdxType="span"
        originalType="span"
        parentName="a"
      />
    </MDXCreateElement>
    Heading 5
  </MDXCreateElement>
  <MDXCreateElement
    id="heading-6"
    mdxType="h6"
    originalType="h6"
  >
    <MDXCreateElement
      aria-hidden="true"
      href="#heading-6"
      mdxType="a"
      originalType="a"
      parentName="h6"
      tabIndex={-1}
    >
      <MDXCreateElement
        className="icon icon-link"
        mdxType="span"
        originalType="span"
        parentName="a"
      />
    </MDXCreateElement>
    Heading 6
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="hr"
    originalType="hr"
  />
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    native markdown image with src relative to root:

    <MDXCreateElement
      alt="pancakes"
      height={452}
      mdxType="img"
      originalType="img"
      parentName="p"
      src="/img/pancakes.jpeg"
      width={900}
    />
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    native markdown image with src relative to root, no leading slash:

    <MDXCreateElement
      alt="pancakes"
      mdxType="img"
      originalType="img"
      parentName="p"
      src="public/img/idontexist.jpeg"
    />
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    native markdown image with src relative to public:

    <MDXCreateElement
      alt="pancakes"
      height={452}
      mdxType="img"
      originalType="img"
      parentName="p"
      src="/img/pancakes.jpeg"
      width={900}
    />
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    native markdown image with src relative to public, no leading slash:

    <MDXCreateElement
      alt="pancakes"
      mdxType="img"
      originalType="img"
      parentName="p"
      src="img/idontexist.jpeg"
    />
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    native markdown image with an absolute src:

    <MDXCreateElement
      alt="G logo"
      mdxType="img"
      originalType="img"
      parentName="p"
      src="https://www.google.com/favicon.ico"
    />
  </MDXCreateElement>
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    JSX image with relative src and size via splatted attrs, others don't magically get sizes:
  </MDXCreateElement>
  <MDXCreateElement
    alt="pancakes"
    height={100}
    mdxType="img"
    originalType="img"
    src="/pancakes.jpeg"
    width={100}
  />
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    JSX image with a hardcoded relative src:
  </MDXCreateElement>
  <MDXCreateElement
    alt="pancakes"
    mdxType="img"
    originalType="img"
    src="/img/pancakes.jpeg"
  />
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    JSX image with a hardcoded relative src and size:
  </MDXCreateElement>
  <MDXCreateElement
    alt="pancakes"
    height={50}
    mdxType="img"
    originalType="img"
    src="/img/pancakes.jpeg"
    width={50}
  />
  <MDXCreateElement
    mdxType="p"
    originalType="p"
  >
    JSX image with an absolute src:
  </MDXCreateElement>
  <MDXCreateElement
    alt="G logo"
    mdxType="img"
    originalType="img"
    src="https://www.google.com/favicon.ico"
  />
</MDXCreateElement>
`;
